syntax = "proto2";
package tdrec.protos;

message FeatureConfig {
    repeated FeatureUnit features = 1;
}

message FeatureUnit {
    oneof feature {
        IdFeature id_feature = 1;
        RawFeature raw_feature = 2;
        SequenceFeature sequence_feature = 3;
        TagFeature tag_feature = 4;
    }
}

message IdFeature {
    required string input_name = 1;
    // feature input, e.g. item:item_id
    optional string feature_name = 2;
    // embedding dimension
    required uint32 embedding_dim = 3;
    // number of id enumerators
    optional uint32 num_buckets = 4;
}

message RawFeature {
    required string input_name = 1;
    // feature input, e.g. item:item_id
    optional string feature_name = 2;
     // raw feature of multiple dimensions
    optional uint32 value_dim = 3 [default = 1];
    // embedding dimension
    optional uint32 embedding_dim = 4 [default = 0];
    //
    optional string separator = 5 [default = "|"];
}

message SequenceFeature {
    required string input_name = 1;
    // feature input, e.g. item:item_id
    optional string feature_name = 2;
    // embedding dimension
    required uint32 embedding_dim = 3;
    // number of id enumerators
    optional uint32 num_buckets = 4;
    //
    optional string separator = 5 [default = "|"];
}

message TagFeature {
    required string input_name = 1;
    // feature input, e.g. item:item_id
    optional string feature_name = 2;
    // embedding dimension
    required uint32 embedding_dim = 3;
    // number of id enumerators
    optional uint32 num_buckets = 4;
    //
    optional string separator = 5 [default = "|"];
}
